libmcount/dynamic.c

```diff
@@ -127,15 +129,25 @@ static int do_dynamic_update(struct symtabs *symtabs, char *patch_funcs)
 				stats.skipped++;
 				break;
 			case 0:
+				success = true;
 			default:
 				break;
 			}
 			stats.total++;
 		}
 
+		if (!success) {
+			nr_failed++;
+			func_name = name;
+		}
+
 		name = strtok(NULL, ";");
 	}
 
+	if (nr_failed)
+		pr_out("%s cannot be patched dynamically\n",
+		       nr_failed > 1 ? "some functions" : func_name);
+
```

Comment:
I tried to use `stats.failed` not using new variable (e.g. `nr_failed`)..

But I think the if statement [(dynamic.c:118~120)](https://github.com/namhyung/uftrace/blob/master/libmcount/dynamic.c#L118-L120) should be considered.
This if statement can just skip it before mcount_patch_func() when given function name was wrong.

```
 87 static int do_dynamic_update(struct symtabs *symtabs, char *patch_funcs)
 88 {
...
117                 for (i = 0; i < symtab->nr_sym; i++) {
118                         sym = &symtab->sym[i];
119 
120                         if ((is_regex && regexec(&re, sym->name, 0, NULL, 0)) ||
121                             (!is_regex && strcmp(name, sym->name)))
122                                 continue;
123 
124                         switch (mcount_patch_func(mdinfo, sym)) {
125                         case -1:
126                                 stats.failed++;
```
For example, even though `unknown_func()` isn't contained in symbol table of a target program (e.g. hello.dyn),
`stats.failed` is 0 zero value because of the if statement [(dynamic.c:118~120)](https://github.com/namhyung/uftrace/blob/master/libmcount/dynamic.c#L118-L120)
```
$ uftrace -v -P unknown_func ./hello.dyn 
uftrace: creating 1 thread(s) for recording
uftrace: using libmcount.so library for tracing
mcount: initializing mcount library
dynamic: dynamic update stats:
dynamic:    total:        0
dynamic:  patched:        0 (0.00%)
dynamic:   failed:        0 (0.00%)
dynamic:  skipped:        0 (0.00%)
mcount: mcount setup done
mcount: shmem_finish: tid: 23718 seqnum = 0 curr = 0, nr_buf = 2 max_buf = 2
unknown_func cannot be patched dynamically
hello
foo
bar
uftrace: child terminated with exit code: 4
uftrace: reading /tmp/uftrace-live-qx7w9H/task.txt file
uftrace: live-record finished.. 
uftrace: start live-replaying...
uftrace: reading /tmp/uftrace-live-qx7w9H/task.txt file
# DURATION    TID     FUNCTION
   1.745 us [23718] | __monstartup();
   1.162 us [23718] | __cxa_atexit();
   2.229 us [23718] | puts();
   0.578 us [23718] | puts();
   0.232 us [23718] | puts();

```